File: backend/src/main/java/se/mistral/backend/chat/ChatService.java
package se.mistral.backend.chat;

import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import se.mistral.backend.chat.dto.ChatMessage;

import java.util.List;

@RequiredArgsConstructor
@Service
public class ChatService {

    private final ChatRepository chatRepository;

    /**
     * Save chat message.
     *
     * @param chatMessage the chat message to save
     * @return the saved chat message
     */
    public ChatMessage saveMessage(ChatMessage chatMessage) {
        Chat chat = new Chat();
        chat.setChatMessage(chatMessage.chatMessage());
        chat.setSenderId(chatMessage.senderId());
        chat.setRecipientId(chatMessage.recipientId());
        chat.setTimestamp(chatMessage.timestamp());
        chatRepository.save(chat);
        return chatMessage;
    }

    /**
     * Retrieve history list.
     *
     * @param senderId    the sender id
     * @param recipientId the recipient id
     * @return the list of all messages between users ordered by timestamp
     */
    public List<ChatMessage> retrieveHistory(long senderId, long recipientId) {
        return chatRepository.findConversationHistory(
                senderId,
                recipientId
        );
    }

}

File: frontend/src/app/core/chat/chat.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';

/**
 * Interface representing a chat message between two users 
 */
export interface ChatMessage {
  senderId: number,
  recipientId: number,
  chatMessage: string,
  timestamp: string
}

@Injectable({ 
  providedIn: 'root' 
})

/**
 * Service for handling chat-related operations, such as retrieving chat history between users.
 * This service uses Angular's HttpClient to communicate with the backend API and fetch chat messages.
 */
export class ChatService {
  private baseUrl = `${environment.apiUrl}/api/chat`;
  private http = inject(HttpClient);

/**
 * Get chat history between two users
 * 
 * @param senderId - ID of the sender
 * @param recipientId - ID of the recipient
 * @returns Observable of an array of ChatMessage objects
 */
  getHistory(senderId: number, recipientId: number): Observable<ChatMessage[]> {
    const params = new HttpParams()
      .set('senderId', senderId)
      .set('recipientId', recipientId);
    return this.http.get<ChatMessage[]>(`${this.baseUrl}/history`, { params });
  }
}